### PR TITLE
New async nbt cache

### PR DESCRIPTION
**Pull request by HookWoods**
The goal of this pr is to reduce I/O operations from the main thread while saving player data and also to avoid too many I/O operations while reading NBT Player file by using a cache (Which start to delete the oldest data when there is too much player compared to the map size)
I don't know if the map implementation is the best method or using a LRU system would be better

### CODE REVIEW

File: patches/server/0066-New-nbt-cache.patch [hunk 1]
```diff
@@ -0,0 +1,137 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Hugo Planque <hookwood01@gmail.com>
+Date: Thu, 21 Jan 2021 17:56:03 +0100
+Subject: [PATCH] New nbt cache
+
```
Comment (Titaniumtown) [inline patches/server/0066-New-nbt-cache.patch]: Maybe put a description about what the patch does in the patch comments here.

**Review by MrIvanPlays — CHANGES_REQUESTED** [review]
yet again _I_ have to make the patch to a level...

**Comment by Titaniumtown** [comment]
> yet again _I_ have to make the patch to a level...

what?

**Comment by MrIvanPlays** [comment]
> > yet again _I_ have to make the patch to a level...
> 
> what?

there are diff problems, missing `// Yatopia` comments, unmarked imports...
I thought hugo would watch me do it on his other 2 prs but looks like he hasn't learned so I'd have to make it _yet again_

**Comment by Titaniumtown** [comment]
Ah ok. I didn't understand what you meant lmao.

**Review by Titaniumtown — CHANGES_REQUESTED** [review]
LGTM, just needs that info in the patch commits, and proper commits when you change a section of code (as @MrIvanPlays said). But either than that, since it's been in a production environment, I see no reason as to why it shouldn't get merged!

**Review by Titaniumtown — APPROVED** [review]
Along with @HookWoods using this patch privately in production for a while, I've been using it for the past 2 days on my server without any issues whatsoever. Seems good to merge into `staging/1.16.5`

File: patches/server/0069-New-nbt-cache.patch [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
+index a1c2bea7c93433434b4e4dfd0bb4b9620657c40d..43487f6c11fcf1e87f4c154da345fc9ab163ff0c 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -824,7 +824,9 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
+         }
+         // Spigot start
+         MCUtil.asyncExecutor.shutdown(); // Paper
++        this.worldNBTStorage.executorService.shutdown();
```
Comment (ishland) [inline patches/server/0069-New-nbt-cache.patch]: // Yatopia

File: patches/server/0069-New-nbt-cache.patch [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
++    public final ExecutorService executorService = Executors.newSingleThreadExecutor(); // Yatopia - NBT Cache system
+ 
+     public WorldNBTStorage(Convertable.ConversionSession convertable_conversionsession, DataFixer datafixer) {
+         this.a = datafixer;
+@@ -29,12 +35,21 @@ public class WorldNBTStorage {
+         try {
+             NBTTagCompound nbttagcompound = entityhuman.save(new NBTTagCompound());
+             File file = File.createTempFile(entityhuman.getUniqueIDString() + "-", ".dat", this.playerDir);
+-
+-            NBTCompressedStreamTools.a(nbttagcompound, file);
```
Comment (ishland) [inline patches/server/0069-New-nbt-cache.patch]: Always comment out deleted code

File: patches/server/0069-New-nbt-cache.patch [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
++            this.executorService.execute(task);
++            // Yatopia end
+         } catch (Exception exception) {
+             WorldNBTStorage.LOGGER.error("Failed to save player data for {}", entityhuman.getName(), exception); // Paper
+         }
+@@ -50,9 +65,17 @@ public class WorldNBTStorage {
+             // Spigot Start
+             boolean usingWrongFile = false;
+             boolean normalFile = file.exists() && file.isFile(); // Akarin - ensures normal file
+-            if ( org.bukkit.Bukkit.getOnlineMode() && !normalFile ) // Paper - Check online mode first // Akarin - ensures normal file
```
Comment (ishland) [inline patches/server/0069-New-nbt-cache.patch]: Always comment out deleted code

File: patches/server/0069-New-nbt-cache.patch [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
++                    playerData = this.dataCache.get(file);
++                }
+                 if ( file.exists() )
+                 {
+                     usingWrongFile = true;
+@@ -61,9 +84,12 @@ public class WorldNBTStorage {
+             }
+             // Spigot End
+ 
+-            if (normalFile) { // Akarin - avoid double I/O operation
```
Comment (ishland) [inline patches/server/0069-New-nbt-cache.patch]: Always comment out deleted code

**Review by ishland — DISMISSED** [review]
Seems good to me but there are still some small issues.

File: patches/server/0069-New-nbt-cache.patch [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
++++ b/src/main/java/net/minecraft/server/WorldNBTStorage.java
+@@ -9,7 +9,11 @@ import org.apache.logging.log4j.Logger;
+ // CraftBukkit start
+ import java.io.FileInputStream;
+ import java.io.InputStream;
++import java.util.concurrent.ExecutorService;
++import java.util.concurrent.Executors;
++
+ import org.bukkit.craftbukkit.entity.CraftPlayer;
++import org.yatopiamc.yatopia.server.cache.NBTCache;
```
Comment (ishland) [inline patches/server/0069-New-nbt-cache.patch]: Don't use imports

**Review by Titaniumtown — APPROVED** [review]
LGTM. I addressed @ishland and my own requests.

**Pull request merged**